src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs

```diff
@@ -56,22 +78,74 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
                     _builder.EmitToken(field, inits.Syntax, _diagnostics.DiagnosticBag);
                     _builder.EmitIntConstant(data.Length);
                     _builder.EmitOpCode(ILOpCode.Cpblk, -3);
-
-                    if (initializationStyle == ArrayInitializerStyle.Mixed)
+                }
+                else
+                {
+                    if (getCreateSpanHelper(_module, elementType) is { } createSpanHelper &&
+                        _module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__GetPinnableReference) is MethodSymbol getPinnableReference)
                     {
-                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
+                        // Use RuntimeHelpers.CreateSpan and cpblk.
+                        EmitStackAllocBlockMultiByteInitializer(data, createSpanHelper, getPinnableReference, elementType, inits.Syntax, _diagnostics.DiagnosticBag);
+                    }
+                    else
+                    {
+                        EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                        mixedInitialized = true;
                     }
                 }
-                else
+
+                if (initializationStyle == ArrayInitializerStyle.Mixed && !mixedInitialized)
                 {
-                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: true);
+                    EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
                 }
             }
+
+            void emitLocalloc()
+            {
+                EmitExpression(count, used: true);
+
+                _sawStackalloc = true;
+                _builder.EmitOpCode(ILOpCode.Localloc);
+            }
+
+            static Cci.IMethodReference? getCreateSpanHelper(Emit.PEModuleBuilder module, TypeSymbol elementType)
+            {
+                var member = module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+                return ((MethodSymbol?)member)?.Construct(elementType).GetCciAdapter();
+            }
+        }
+
+        private void EmitStackAllocBlockMultiByteInitializer(ImmutableArray<byte> data, Cci.IMethodReference createSpanHelper, MethodSymbol getPinnableReferenceDefinition, TypeSymbol elementType, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        {
+            var readOnlySpan = getPinnableReferenceDefinition.ContainingType.Construct(elementType);
+            Debug.Assert(TypeSymbol.Equals(readOnlySpan.OriginalDefinition, _module.Compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T), TypeCompareKind.ConsiderEverything));
+            var getPinnableReference = getPinnableReferenceDefinition.AsMember(readOnlySpan);
+
+            _builder.EmitOpCode(ILOpCode.Dup);
```

Comment:
>_builder.EmitOpCode(ILOpCode.Dup); [](http://example.com/codeflow?start=12&length=34)

What is it we are duplicating? It would be good to add a comment what this helper assumes to have on the stack. Alternatively, consider converting this helper to a local function, or even inlining the code. It is used only in one place.